### PR TITLE
Refactor state system.

### PR DESCRIPTION
**Pull request by MeKeyCool**

### CODE REVIEW

File: src/public/client.js [hunk 1]
```diff
@@ -78,37 +121,43 @@
...
       else button.classList.remove('selected');
     });
-
-    controller.selectedOption = controller.game.options[selectedOption];
   }
 
   function selectOption(event) {
     var optionIndex = event.target.getAttribute('data-option-index');
-    selectedOption = optionIndex;
+    controller.selectedOption = parseInt(optionIndex);
```
Comment (silverspectro) [inline src/public/client.js:134]: missing the decimal operator for the `parseInt`.
second argument.
```
parseInt(optionIndex, 10);
```

File: src/public/client.js [hunk 2]
```diff
@@ -78,37 +121,43 @@
...
 
   function updateGameOptions(options) {
+
+    if(options.length <= 0) {
+      throw new Error('Invalid option list for controll.')
+    }
+
     wipeElementsFrom(optionListUl);
-    options.forEach(function (option, index) {
+    if(options.indexOf(controller.selectedOption) == -1) {
```
Comment (silverspectro) [inline src/public/client.js:145]: `===` not `==`

File: src/public/server.js [hunk 3]
```diff
@@ -183,10 +183,20 @@ Dungeon.prototype = {
     });
 
     this.socket.on('join-game', function (payload) {
-      var controller = find(controllers, payload.gameId);
-      var game = controller ? controller.game : undefined;
-      self.createArea(payload.areaWidth, payload.areaHeight);
-      if (game && !game.started) self.joinRoom(game);
+      var syncId = payload.gameId;
```
Comment (silverspectro) [inline src/public/server.js:186]: why `syncId` ?
It's a notion we never have in the application, i find it rather confusing.
I think `gameId` is more appropriate and understandable.

File: src/public/server.js [hunk 4]
```diff
@@ -223,195 +236,121 @@ Dungeon.prototype = {
...
-      case 'trap':
-        {
-          if (victim.area[x][y].state === 'square') {
-            if (this.money >= this.config.trapCost) moneyToDeduce += this.config.trapCost;
-          }
-          break;
-        }
+  // This method should not be called if player doesn't have enough money
+  deduceMoney: function (requestedState) {
+    /// @TODO make a global cell state object to centralize label, cost, etc.
```
Comment (silverspectro) [inline src/public/server.js:258]: this TODO seems done.
As the config for the costs are in the dungeons and the global state variables are done.
maybe remove ?

File: src/public/server.js [hunk 5]
```diff
@@ -420,34 +359,21 @@ Dungeon.prototype = {
...
-          },
-          state: squareStates[0],
-        });
-      }
-    }
+  createArea: function (columns, rows) {
+    this.area.reset(columns, rows);
+    
+    // set player position
+    var playerXPos = Math.floor( (columns-1) / 2);
```
Comment (silverspectro) [inline src/public/server.js:366]: maybe a comment for why we need the `-1` ?

**Review by silverspectro — APPROVED** [review]

File: src/public/client.js (hunk 2 above)
Comment (MeKeyCool) [reply src/public/client.js:145]: I thought the difference was due to implicit transtyping and i don't know anything that could be implicitly transtyped to `-1`.
Did I missed something ?

File: src/public/server.js (hunk 3 above)
Comment (MeKeyCool) [reply src/public/server.js:186]: I called it syncId cause we use this id to get controller as well as corresponding dungeon. I understood this id as a way to associate objects and i felt this name more expressive.

File: src/public/server.js (hunk 4 above)
Comment (MeKeyCool) [reply src/public/server.js:258]: Yes, as we discussed a bit, i am not sure about this idea anymore. I won't do a commit just to remove this comment but i'll consider it in coming developments.

**Pull request merged**